toolkit/cli/smart-contracts-commands/src/permissioned_candidates.rs

```diff
@@ -0,0 +1,64 @@
+use crate::PaymentFilePath;
+use ogmios_client::jsonrpsee::client_for_url;
+use partner_chains_cardano_offchain::await_tx::FixedDelayRetries;
+use partner_chains_cardano_offchain::permissioned_candidates::upsert_permissioned_candidates;
+use sidechain_domain::AuraPublicKey;
+use sidechain_domain::GrandpaPublicKey;
+use sidechain_domain::PermissionedCandidateData;
+use sidechain_domain::SidechainPublicKey;
+use sidechain_domain::UtxoId;
+
+#[derive(Clone, Debug, clap::Parser)]
+pub struct UpsertPermissionedCandidatesCmd {
+	#[clap(flatten)]
+	common_arguments: crate::CommonArguments,
+	#[arg(long)]
+	permissioned_candidates_file: String,
+	#[clap(flatten)]
+	payment_key_file: PaymentFilePath,
+	#[arg(long, short('c'))]
+	genesis_utxo: UtxoId,
+}
+
+impl UpsertPermissionedCandidatesCmd {
+	pub async fn execute(self) -> crate::CmdResult<()> {
+		let payment_key = self.payment_key_file.read_key()?;
+
+		let mut reader = csv::ReaderBuilder::new()
+			.delimiter(b':')
```

Comment:
Could you share a file that you have tested this with? I tried a CSV that seems valid, but I get empty result.

Reply:
With 1 entry file this code reads 0 records, with 3 entries in file it reads 2 records.